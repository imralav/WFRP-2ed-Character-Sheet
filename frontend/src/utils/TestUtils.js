import React, {Fragment} from 'react';

import en from 'react-intl/locale-data/en';
import {addLocaleData, IntlProvider} from 'react-intl';
import renderer from 'react-test-renderer';

import {BrowserRouter as Router, Route} from 'react-router-dom';

addLocaleData(en);

class ComponentBuilder {
    constructor(component) {
        this.component = component;
    }

    withIntl() {
        this.component = (
            <IntlProvider locale="en" messages={{}}>
                <Fragment>
                {this.component}
                </Fragment>
            </IntlProvider>
        );
        return this;
    }

    withRouter() {
        this.component = (
            <Router>
                <Fragment>
                    {this.component}
                </Fragment>
            </Router>
        );
        return this;
    }

    build() {
        return renderer.create(this.component);
    }
}

export default class TestUtils {
    static buildComponent(component) {
        return new ComponentBuilder(component);
    }
}