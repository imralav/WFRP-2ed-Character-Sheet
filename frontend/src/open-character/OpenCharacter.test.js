import React from 'react';
import OpenCharacter from '../open-character';
import TestUtils from '../utils/TestUtils';

describe('OpenCharacter', () => {
    const CHARACTER_DATA = {player: {name: 'playerName'}, character: {name: 'characterName'}};
    const CHARACTER_DATA_JSON = JSON.stringify(CHARACTER_DATA);
    const ENCODED_CHARACTER_DATA = btoa(CHARACTER_DATA_JSON);
    const CORRUPTED_CHARACTER_DATA = 'someCorruptedData';
    const characterLoadValidator = {isValid: jest.fn()};
    let component;
    let componentRoot;
    let tree;
    beforeEach(() => {
        component = TestUtils.buildComponent(
            <OpenCharacter characterLoadValidator={characterLoadValidator}/>
        ).withIntl().build();
        componentRoot = component.root;
        tree = component.toJSON();
    });
    it('should match previous snapshot', () => {
        expect(tree).toMatchSnapshot();
    });
    describe('uncorrupted character data change', () => {
        beforeEach(() => {
            componentRoot.findByType('textarea').props.onChange({
                target: {
                    value: ENCODED_CHARACTER_DATA
                }
            });
        });
        it('should add unparsed character data to state', () => {
            expect(componentRoot.findByType(OpenCharacter).instance.state.unparsedCharacterData).toEqual(ENCODED_CHARACTER_DATA);
        });
        describe('loadCharacterData', () => {
            describe('on passed validation character data load', () => {
                beforeEach(() => {
                    characterLoadValidator.isValid.mockImplementation(() => true);
                    componentRoot.findByType('button').props.onClick();
                });
                it('should parse character data and set it on state', () => {
                    expect(componentRoot.findByType(OpenCharacter).instance.state.characterData).toEqual(CHARACTER_DATA);
                });
                it('should match snapshot', () => {
                    expect(component.toJSON()).toMatchSnapshot();
                });
            });
            describe('on failed validation character data load', () => {
                beforeEach(() => {
                    characterLoadValidator.isValid.mockImplementation(() => false);
                    componentRoot.findByType('button').props.onClick();
                });
                it('should not parse character data and set it on state', () => {
                    expect(componentRoot.findByType(OpenCharacter).instance.state.characterData).not.toBeDefined();
                });
                it('should match snapshot', () => {
                    expect(component.toJSON()).toMatchSnapshot();
                });
            });
        });
    });
    describe('corrupted character data change', () => {
        beforeEach(() => {
            componentRoot.findByType('textarea').props.onChange({
                target: {
                    value: CORRUPTED_CHARACTER_DATA
                }
            });
        });
        it('should add unparsed character data to state', () => {
            expect(componentRoot.findByType(OpenCharacter).instance.state.unparsedCharacterData).toEqual(CORRUPTED_CHARACTER_DATA);
        });
        describe('loadCharacterData', () => {
            describe('on decoding errors', () => {
                beforeEach(() => {
                    componentRoot.findByType('button').props.onClick();
                });
                it('should not parse character data and set it on state', () => {
                    expect(componentRoot.findByType(OpenCharacter).instance.state.characterData).not.toBeDefined();
                });
                it('should set invalid state', () => {
                    expect(componentRoot.findByType(OpenCharacter).instance.state.isValid).toEqual(false);
                });
                it('should match snapshot', () => {
                    expect(component.toJSON()).toMatchSnapshot();
                });
            });
        });
    });
});
