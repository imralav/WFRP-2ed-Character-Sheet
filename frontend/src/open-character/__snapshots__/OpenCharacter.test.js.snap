// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenCharacter corrupted character data change loadCharacterData on decoding errors should match snapshot 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-12 offset-sm-2 col-sm-8"
    >
      <label
        className="label-font-size"
      >
        <span>
          Paste character data below
        </span>
      </label>
      <div
        className="text-danger"
      >
        <span>
          Provided character data is incorrect. Make sure to paste the exact same text as was provided by the application.
        </span>
      </div>
      <textarea
        className="content-font-size"
        id="character-data-input"
        onChange={[Function]}
        onFocus={[Function]}
      />
      <button
        className="btn btn-dark btn-sm content-font-size"
        onClick={[Function]}
      >
        <span>
          Open pasted character
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`OpenCharacter should match previous snapshot 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-12 offset-sm-2 col-sm-8"
    >
      <label
        className="label-font-size"
      >
        <span>
          Paste character data below
        </span>
      </label>
      
      <textarea
        className="content-font-size"
        id="character-data-input"
        onChange={[Function]}
        onFocus={[Function]}
      />
      <button
        className="btn btn-dark btn-sm content-font-size"
        onClick={[Function]}
      >
        <span>
          Open pasted character
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`OpenCharacter uncorrupted character data change loadCharacterData on failed validation character data load should match snapshot 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-12 offset-sm-2 col-sm-8"
    >
      <label
        className="label-font-size"
      >
        <span>
          Paste character data below
        </span>
      </label>
      <div
        className="text-danger"
      >
        <span>
          Provided character data is incorrect. Make sure to paste the exact same text as was provided by the application.
        </span>
      </div>
      <textarea
        className="content-font-size"
        id="character-data-input"
        onChange={[Function]}
        onFocus={[Function]}
      />
      <button
        className="btn btn-dark btn-sm content-font-size"
        onClick={[Function]}
      >
        <span>
          Open pasted character
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`OpenCharacter uncorrupted character data change loadCharacterData on passed validation character data load should match snapshot 1`] = `
<div
  className="row character-sheet"
>
  <div
    className="col-12"
    id="character-json"
  >
    <button
      className="btn btn-dark btn-sm content-font-size"
      onClick={[Function]}
    >
      <span>
        Copy character data to clipboard for later use
      </span>
    </button>
  </div>
  <div
    className="col-12"
    id="wfrp-logo-container"
  >
    <img
      alt="WFRP Logo"
      src="wfrp-logo.png"
    />
  </div>
  <div
    className="col-12"
    id="player"
  >
    <table>
      <thead>
        <tr>
          <th
            className="header-font-size"
            colSpan="2"
          >
            <span>
              Player
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Name
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="name"
                onChange={[Function]}
                type="text"
                value="playerName"
              />
            </div>
          </td>
          <td
            className="w-40p"
          >
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Game Master
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="gameMaster"
                onChange={[Function]}
                type="text"
                value={undefined}
              />
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Campaign
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="campaign"
                onChange={[Function]}
                type="text"
                value={undefined}
              />
            </div>
          </td>
          <td>
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Campaign Year
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="campaignYear"
                onChange={[Function]}
                type="text"
                value={undefined}
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="col-12"
    id="character"
  >
    <table>
      <thead>
        <tr>
          <th
            className="header-font-size"
          >
            <span>
              Character
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Name
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="name"
                onChange={[Function]}
                type="text"
                value="characterName"
              />
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Race
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="race"
                onChange={[Function]}
                type="text"
                value={undefined}
              />
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Current Career
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="currentCareer"
                onChange={[Function]}
                type="text"
                value={undefined}
              />
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Previous Careers
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="previousCareers"
                onChange={[Function]}
                type="text"
                value={undefined}
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="col-12"
    id="experience-points"
  >
    <table>
      <thead>
        <tr>
          <th
            className="header-font-size"
            colSpan="2"
          >
            <span>
              Experience Points
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Current
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="current"
                onChange={[Function]}
                type="number"
                value={0}
              />
            </div>
          </td>
          <td>
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Total
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="total"
                onChange={[Function]}
                type="number"
                value={0}
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="col-12"
    id="personal-details"
  >
    <table>
      <thead>
        <tr>
          <th
            className="header-font-size"
            colSpan="2"
          >
            <span>
              Personal Details
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Age
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="age"
                onChange={[Function]}
                type="number"
                value={0}
              />
            </div>
          </td>
          <td>
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Gender
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="gender"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Eye Color
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="eyeColor"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
          </td>
          <td>
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Weight
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="weight"
                onChange={[Function]}
                type="number"
                value={0}
              />
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Hair Color
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="hairColor"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
          </td>
          <td>
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Height
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="height"
                onChange={[Function]}
                type="number"
                value={0}
              />
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Star Sign
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="starSign"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
          </td>
          <td>
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Number of Siblings
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="numberOfSiblings"
                onChange={[Function]}
                type="number"
                value={0}
              />
            </div>
          </td>
        </tr>
        <tr>
          <td
            colSpan="2"
          >
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Birthplace
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="birthplace"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
          </td>
        </tr>
        <tr>
          <td
            colSpan="2"
          >
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Distinguishing Marks
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="distinguishingMarks"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="col-12"
    id="character-profile"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Character Profile
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="combat-movement"
  >
    <table>
      <thead>
        <tr>
          <th
            className="header-font-size"
            colSpan="3"
          >
            <span>
              Combat Movement
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            className="w-33p"
          >
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Move/Disengage
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="moveDisengage"
                onChange={[Function]}
                type="number"
                value={0}
              />
            </div>
          </td>
          <td
            className="w-33p"
          >
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Charge Attack
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="chargeAttack"
                onChange={[Function]}
                type="number"
                value={0}
              />
            </div>
          </td>
          <td
            className="w-33p"
          >
            <div>
              <label
                className="label-font-size"
              >
                <span>
                  Run
                </span>
                :
              </label>
              <input
                className="content-font-size"
                name="run"
                onChange={[Function]}
                type="number"
                value={0}
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="col-12"
    id="weapons"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Weapons
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="armour"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Armour
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="armour-points"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Armour Points
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="skills"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Skills
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="talents"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Talents
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="trappings"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Trappings
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="money"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Money
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <span>
              Gold Crowns (gc)
            </span>
            :
          </td>
        </tr>
        <tr>
          <td>
            <span>
              Silver Schillings (s)
            </span>
            :
          </td>
        </tr>
        <tr>
          <td>
            <span>
              Brass Pennies (p)
            </span>
            :
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
